keps/0013-pull-based-audit-events-stream.md

```diff
@@ -0,0 +1,199 @@
+       // groups in the audit event, if all specified groups exist in the audit event, then it will be sent to user.
+       // optional
+       groups []string
+       // namespace of the audit event, <none> for cluster scoped request, empty string for all namespaces.
+       // only supported for /audits endpoint.
+       // optional
+       namespace string
+       // apiGroup in the audit event, <core> for the core api group.
+       // optional
+       apiGroup string
+       // resource in the audit event, for example: pods.
+       // optional
+       resource string
+       // verb in the audit event, for example: create, put. If the verbs contains the verb of audit event, then
+       // the audit event will be sent to user.
+       // optional
+       verbs []string
+}
+```
+
+### Bandwidth limit Configuration
+```golang
+// LimitType is the type of the limit.
+type LimitType string
+
+const (
+	// ServerLimitType is a type of limit where there is one bucket shared by
+	// all of the audit events sent by the API Server, including '/audits' and
+	// 'audits/{namespace}' endpoints.
+	ServerLimitType LimitType = "server"
+
+	// NamespaceLimitType is a type of limit where there is one bucket used by
+	// each namespace. When the namespace is set to empty string, it limit the
+	// bandwidth of all requests to '/audits' endpoint.
+	NamespaceLimitType LimitType = "namespace"
+)
+
+// Configuration provides bandwidth configuration for the pull based audit event
+// endpoints.
+type Configuration struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// limits are the limits to place on audit events sent by API server or
+	// each endpoint.
+	Limits []Limit `json:"limits"`
+}
+
+
+// Limit is the configuration for a particular limit type
+type Limit struct {
+	// type is the type of limit to which this configuration applies
+	Type LimitType `json:"type"`
+
+	// ThrottleQPS is the number of bytes sent per second that are allowed for this
+	// type of limit.
+	ThrottleQPS int32 `json:"throttleQps"`
+
+	// ThrottleBurst defines the maximum number of bytes sent to the client at the
+	// same moment.
+	ThrottleBurst int32 `json:"throttleBurst"`
+
+	// MaxEventSize defines max allowed size of the event. If the event is larger, truncating will be performed
+	// before emitting to client.
+	// optional
+	MaxEventSize int64 `json:"maxEventSize"`
+}
+```
+
+When the sent bytes exceeds the bandwidth, API server will try to truncate the audit event, and check the limit again. If bandwidth limit is still exceeded, the event would be droped.
+
+
+### User Stories
+
+#### Story 1
+As a tenant, I will be able to known all things happened to my own namespace. When problem happens, I can use the audit events for trouble shooting.
+
+#### Story 2
+As a developer, I will easily be able to get a subset of audit events. This will save a lot of time for debugging. I don't need to grep audit events from a couple of audit files.
+
+### Risks and Mitigations
```

Comment:
Would be good to talk about SLO. In particular, I think these endpoints should be treated as debugging & development endpoints, not for actual secure audit logging.

Reply:
Will add it.